### PR TITLE
modify gluon image_classification for imagenet

### PR DESCRIPTION
**Pull request by zhreshold**
make gluon example able to train imagenet

improve symbolic

add std option to recordio_2.cc/recordio.cc

### CODE REVIEW

**Comment by piiswrong** [comment]
Please rebase to latest master

File: example/gluon/image_classification.py [hunk 1]
```diff
@@ -165,8 +176,16 @@ def train(epochs, ctx):
...
-        mod.fit(train_data, num_epoch=opt.epochs, kvstore=opt.kvstore,
-                batch_end_callback = mx.callback.Speedometer(batch_size, 1))
+        mod.fit(train_data,
+                eval_data = val_data,
+                num_epoch=opt.epochs,
+                kvstore=opt.kvstore,
+                batch_end_callback = mx.callback.Speedometer(batch_size, max(1, opt.log_interval)),
+                epoch_end_callback = mx.callback.do_checkpoint('image-classifier-%s'% opt.model),
+                optimizer = 'sgd',
+                optimizer_params = {'learning_rate': opt.lr, 'wd': opt.wd},
```
Comment (piiswrong) [inline example/gluon/image_classification.py]: momentum？

File: src/io/iter_image_recordio_2.cc [hunk 2]
```diff
@@ -388,44 +406,23 @@ inline void ImageRecordIOParser2<DType>::ParseChunk(dmlc::InputSplit::Blob * chu
...
-            } else if (!meanfile_ready_ || normalize_param_.mean_img.length() == 0) {
-              // do not subtract anything
-              for (int k = 0; k < n_channels; ++k) {
-                RGBA[k] = RGBA[k] * normalize_param_.scale;
-              }
-            } else {
-              CHECK(meanfile_ready_);
-              for (int k = 0; k < n_channels; ++k) {
-                  RGBA[k] = (RGBA[k] - meanimg_[k][i][j]) * contrast_scaled + illumination_scaled;
+            for (int k = 0; k < n_channels; ++k) {
```
Comment (piiswrong) [inline src/io/iter_image_recordio_2.cc]: Is this the same as before? Looks like normalize_param_.scale was only used for meanfile before?

File: src/io/iter_image_recordio_2.cc [hunk 3]
```diff
@@ -388,44 +406,23 @@ inline void ImageRecordIOParser2<DType>::ParseChunk(dmlc::InputSplit::Blob * chu
...
+              if (meanfile_ready_) {
+                RGBA[k] = (RGBA[k] - meanimg_[k][i][j]) * RGBA_MULT[k] + RGBA_BIAS[k];
+              } else if (RGBA_MEAN[k] != 0.0f) {
```
Comment (piiswrong) [inline src/io/iter_image_recordio_2.cc]: seems redundant to check for != 0

File: src/io/iter_image_recordio_2.cc (hunk 2 above)
Comment (zhreshold) [reply src/io/iter_image_recordio_2.cc]: The previous one is wrong, if meanfile_ is ready, the normalize_param_.scale was not multiplied.

File: src/io/iter_image_recordio_2.cc [hunk 4]
```diff
@@ -379,6 +379,24 @@ inline void ImageRecordIOParser2<DType>::ParseChunk(dmlc::InputSplit::Blob * chu
           - normalize_param_.max_random_illumination) * normalize_param_.scale;
```
Comment (piiswrong) [inline src/io/iter_image_recordio_2.cc]: normalize_param_.scale is multiplied twice?

Comment (zhreshold) [reply src/io/iter_image_recordio_2.cc]: my bad, mislead by the scale when mean is 0, should remove that.

File: src/io/iter_image_recordio_2.cc [hunk 5]
```diff
@@ -379,6 +379,24 @@ inline void ImageRecordIOParser2<DType>::ParseChunk(dmlc::InputSplit::Blob * chu
...
       }
       DType RGBA[4] = {};
+      float RGBA_MULT[4] = {
```
Comment (piiswrong) [inline src/io/iter_image_recordio_2.cc]: Should this also be excluded when DType is uint8?

**Comment by szha** [comment]
Thanks for fixing the example, could you also change the `print(opt)` to `logging.info(opt)` near line 71?

File: src/io/iter_normalize.h [hunk 6]
```diff
@@ -144,35 +148,71 @@ class ImageNormalizeIter : public IIterator<DataInst> {
...
-        data[3] -= param_.mean_a;
+    // setup mean if not loaded from file, should happen only once
+    if (!meanfile_ready_ || (meanimg_.shape_ != data.shape_)) {
+      if (param_.mean_r > 0.0f || param_.mean_g > 0.0f ||
+          param_.mean_b > 0.0f || param_.mean_a > 0.0f) {
+        // setup meanimg using mean_r/g/b/a
+        meanimg_.Resize(data.shape_);
+        switch (static_cast<int>(data.shape_[0])) {
+          case 4: {
+            meanimg_[3] = param_.mean_a;
```
Comment (piiswrong) [inline src/io/iter_normalize.h]: why modify meanimg and stdimg? Feels like a bad idea

Comment (zhreshold) [reply src/io/iter_normalize.h]: To condense operation and avoid too many branches for each input.
the meanimg_ and stdimg_ should be modified only once for the first image. I would like it to happen in Init, but can't infer the shape there.

Comment (piiswrong) [reply src/io/iter_normalize.h]: But reshaping and re filling meanimg on each iter is slow

**Comment by zhreshold** [comment]
@piiswrong Could you review again?
